### PR TITLE
ci: Switch to Ubuntu 24.04

### PR DESCRIPTION
**Pull request by daandemeyer**
Let's see what breaks

### CODE REVIEW

**Comment by bluca** [comment]
> Let's see what breaks

Looks like "everything" is the answer

**Comment by bluca** [comment]
This looks worse than usual:

```
30/92 systemd:integration-tests / TEST-01-BASIC                                                 FAIL              5.93s   exit status 1
36/92 systemd:integration-tests / TEST-04-JOURNAL                                               FAIL            195.86s   exit status 1
38/92 systemd:integration-tests / TEST-07-PID1                                                  FAIL             74.52s   exit status 1
56/92 systemd:integration-tests / TEST-35-LOGIN                                                 FAIL              9.32s   exit status 1
58/92 systemd:integration-tests / TEST-43-PRIVATEUSER-UNPRIV                                    FAIL              7.38s   exit status 1
61/92 systemd:integration-tests / TEST-45-TIMEDATE                                              FAIL             10.79s   exit status 1
73/92 systemd:integration-tests / TEST-65-ANALYZE                                               FAIL              7.85s   exit status 1
81/92 systemd:integration-tests / TEST-74-AUX-UTILS                                             FAIL              9.50s   exit status 1
88/92 systemd:integration-tests / TEST-81-GENERATORS                                            FAIL              9.30s   exit status 1
```

**Review by bluca — APPROVED** [review]

File: .github/workflows/cifuzz.yml [hunk 1]
```diff
@@ -22,7 +22,8 @@ on:
       - main
 jobs:
   Fuzzing:
-    runs-on: ubuntu-latest
+    # FIXME: Figure out why 32-bit applications fail to run in docker on Ubuntu 24.04.
+    runs-on: ubuntu-22.04
```
Comment (evverx) [inline .github/workflows/cifuzz.yml:26]: `cflite_pr.yml` uses basically the same OSS-Fuzz base-builder images and the same runtime so `runs-on` should probably be set to `ubuntu-22.04` there too.

File: .github/workflows/build_test.sh [hunk 2]
```diff
@@ -84,6 +84,14 @@ if [[ "$COMPILER" == clang ]]; then
     CXX="clang++-$COMPILER_VERSION"
     AR="llvm-ar-$COMPILER_VERSION"
 
+    if systemd-analyze compare-versions "$COMPILER_VERSION" ge 17; then
+        CFLAGS="-fno-sanitize=function"
+        CXXFLAGS="-fno-sanitize=function"
```
Comment (evverx) [inline .github/workflows/build_test.sh:89]: I don't think systemd is built with UBSan in this workflow so I'm not sure why this is needed. If it's necessary I think it would be better to pass `c_args` and so on so that it was consistent.

Comment (daandemeyer) [reply .github/workflows/build_test.sh:89]: If I remember correctly it was building the fuzz tests that broke without this.

Comment (evverx) [reply .github/workflows/build_test.sh:89]: It should be fixed in `meson.build` responsible for building those fuzz targets then.

File: .github/workflows/cifuzz.yml (hunk 1 above)
Comment (evverx) [reply .github/workflows/cifuzz.yml:26]: I opened https://github.com/google/oss-fuzz/issues/12042

File: .github/workflows/build_test.sh (hunk 2 above)
Comment (evverx) [reply .github/workflows/build_test.sh:89]: Either way that action switched to clang-18 in 4651e1428dae935bb32dec082db5155f2e8b1c98 and systemd didn't fail to compile so I have to admit I'm not sure why it would fail here.

File: .github/workflows/linter.yml [hunk 3]
```diff
@@ -29,7 +29,7 @@ jobs:
           fetch-depth: 0
 
       - name: Lint Code Base
-        uses: super-linter/super-linter/slim@4e51915f4a812abf59fed160bb14595c0a38a9e7
+        uses: super-linter/super-linter/slim@88ea3923a7e1f89dd485d079f6eb5f5e8f937589
```
Comment (evverx) [inline .github/workflows/linter.yml:32]: I'm kind of confused. Why was it bumped here? It doesn't seem to be related to switching to Ubuntu 24.04.

Comment (evverx) [reply .github/workflows/linter.yml:32]: I'm guessing it fixed:
```
  proxy | 2024/06/05 07:32:38 [098] 200 https://github.com:443/super-linter/super-linter/git-upload-pack
updater | 2024/06/05 07:32:43 ERROR <job_837719984> Error processing super-linter/super-linter (Dependabot::SharedHelpers::HelperSubprocessFailed)
updater | 2024/06/05 07:32:43 ERROR <job_837719984> error: no such commit 4e51915f4a812abf59fed160bb14595c0a38a9e7
updater | 2024/06/05 07:32:43 ERROR <job_837719984> /home/dependabot/common/lib/dependabot/shared_helpers.rb:432:in `run_shell_command'
```
but it is unrelated to Ubuntu 24.04 anyway.

Looking at https://github.com/evverx/avahi/actions/runs/9437349361/job/25993098340
```
  .github/workflows/cifuzz.yml:22:14: label "ubuntu-24.04" is unknown. available labels are "windows-latest", "windows-latest-8-cores", "windows-2022", "windows-2019", "windows-2016", "ubuntu-latest", "ubuntu-latest-4-cores", "ubuntu-latest-8-cores", "ubuntu-latest-16-cores", "ubuntu-22.04", "ubuntu-20.04", "macos-latest", "macos-latest-xl", "macos-latest-xlarge", "macos-latest-large", "macos-14-xl", "macos-14-xlarge", "macos-14-large", "macos-14", "macos-14.0", "macos-13-xl", "macos-13-xlarge", "macos-13-large", "macos-13", "macos-13.0", "macos-12-xl", "macos-12-xlarge", "macos-12-large", "macos-12", "macos-12.0", "macos-11", "macos-11.0", "macos-10.15", "self-hosted", "x64", "arm", "arm64", "linux", "macos", "windows". if it is a custom label for self-hosted runner, set list of labels in actionlint.yaml config file [runner-label]
     |
  22 |     runs-on: ubuntu-24.04
     |              ^~~~~~~~~~~~
  ------
```
looks like it's related to Ubuntu 24.04 after all. It would be nice to mention things like that in commit messages though.

File: .github/workflows/build_test.sh (hunk 2 above)
Comment (evverx) [reply .github/workflows/build_test.sh:89]: I was curious so I reverted the CFLAGS part in my fork and pushed it to see what can possibly fail. All the build jobs are green: https://github.com/evverx/systemd/actions/runs/9435948677/job/25990022282?pr=114.

File: .github/workflows/cifuzz.yml (hunk 1 above)
Comment (evverx) [reply .github/workflows/cifuzz.yml:26]: It's probably https://github.com/actions/runner-images/issues/9977#issuecomment-2153044449
